### PR TITLE
Fixed #12603 Errors logged after cross context dispatch

### LINKED ISSUE #12603: ee9 / UnsupportedOperationException: Read Only
**Jetty version(s)**
12.0.x HEAD 

**Jetty Environment**
ee9 (but might exist in others too)

**Java version/vendor** `(use: java -version)`
Adoptium 21

**OS type/version**
Linux

**Description**
An exception seen during build ...

```
2024-12-03 07:01:38.656:WARN :oejen.HttpChannel:qtp1727361096-31: /context/resourceServlet/content.txt
java.lang.UnsupportedOperationException: Read Only
	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.internal.ResponseHttpFields$1.remove(ResponseHttpFields.java:153)
	at org.eclipse.jetty.ee9.nested@12.0.15-SNAPSHOT/org.eclipse.jetty.ee9.nested.Response.resetContent(Response.java:1249)
	at org.eclipse.jetty.ee9.nested@12.0.15-SNAPSHOT/org.eclipse.jetty.ee9.nested.HttpChannel.handle(HttpChannel.java:532)
	at org.eclipse.jetty.ee9.nested@12.0.15-SNAPSHOT/org.eclipse.jetty.ee9.nested.ContextHandler$CoreContextHandler$CoreToNestedHandler.handle(ContextHandler.java:2873)
	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1060)
	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:181)
	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.Server.handle(Server.java:182)
	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:662)
	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:418)
	at org.eclipse.jetty.io@12.0.15-SNAPSHOT/org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io@12.0.15-SNAPSHOT/org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io@12.0.15-SNAPSHOT/org.eclipse.jetty.io.ByteArrayEndPoint.lambda$new$0(ByteArrayEndPoint.java:58)
	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Suppressed: 
	|java.lang.UnsupportedOperationException: Read Only
	|	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.internal.ResponseHttpFields$1.remove(ResponseHttpFields.java:153)
	|	at org.eclipse.jetty.ee9.nested@12.0.15-SNAPSHOT/org.eclipse.jetty.ee9.nested.Response.resetContent(Response.java:1249)
	|	at org.eclipse.jetty.ee9.nested@12.0.15-SNAPSHOT/org.eclipse.jetty.ee9.nested.HttpChannel.handle(HttpChannel.java:566)
	|	at org.eclipse.jetty.ee9.nested@12.0.15-SNAPSHOT/org.eclipse.jetty.ee9.nested.ContextHandler$CoreContextHandler$CoreToNestedHandler.handle(ContextHandler.java:2873)
	|	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1060)
	|	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:181)
	|	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.Server.handle(Server.java:182)
	|	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:662)
	|	at org.eclipse.jetty.server@12.0.15-SNAPSHOT/org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:418)
	|	at org.eclipse.jetty.io@12.0.15-SNAPSHOT/org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	|	at org.eclipse.jetty.io@12.0.15-SNAPSHOT/org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	|	at org.eclipse.jetty.io@12.0.15-SNAPSHOT/org.eclipse.jetty.io.ByteArrayEndPoint.lambda$new$0(ByteArrayEndPoint.java:58)
	|	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	|	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	|	at org.eclipse.jetty.util/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	|	at java.base/java.lang.Thread.run(Thread.java:1570)
```

Edit: gregw - confirmed it is in 12.0.x

### PR DESCRIPTION
**Pull request by gregw**
Fixed #12603 Errors logged after cross context dispatch:
 + Update bytes written in HttpOutput when bypassed by optimization
 + abort if error dispatch throws

### CODE REVIEW

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java [hunk 1]
```diff
@@ -333,7 +336,10 @@ public Mutable add(HttpHeader header, String value)
         @Override
         public Mutable add(HttpField field)
         {
-            _response.addHeader(field.getName(), field.getValue());
+            if (field.getHeader() == HttpHeader.CONTENT_LENGTH)
+                _response.setContentLengthLong(field.getLongValue());
+            else
+                _response.addHeader(field.getName(), field.getValue());
```
Comment (joakime) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java:342]:
```suggestion
            if (field.getHeader() == HttpHeader.CONTENT_LENGTH)
                _response.setContentLengthLong(field.getLongValue());
            _response.addHeader(field.getName(), field.getValue());
```

In the Content-Length specific case, we don't want to drop the normal response.addHeader too.  right?

Comment (gregw) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java:342]: @joakime The setContentLengthLong does the addHeader, but with a long value included, so we don't have to parse the string to work out the content length.

**Comment by gregw** [comment]
@sbordet @janbartel review please!

**Review by joakime — DISMISSED** [review]

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpChannel.java [hunk 2]
```diff
@@ -561,16 +560,7 @@ public boolean handle()
...
-                            {
-                                _response.resetContent();
-                                sendResponseAndComplete();
-                            }
-                            catch (Throwable t)
-                            {
-                                ExceptionUtil.addSuppressedIfNotAssociated(x, t);
-                                throw x;
-                            }
+                            abort(x);
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpChannel.java:563]: The logic is now completely different from before... Is that ok?

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java [hunk 3]
```diff
@@ -137,12 +137,15 @@ public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
             last = false;
         try
         {
-            if (!_wrapped && !_baseResponse.isWritingOrStreaming())
+            if (last && !_wrapped && !_baseResponse.isWritingOrStreaming())
             {
+                // We can bypass the HttpOutput stream, but we need to update its bytes written
+                _baseResponse.getHttpOutput().setBytesWritten(byteBuffer.remaining());
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java]: If the write at the line below fails, the bytes would have been updated, but they were not written.

`HttpOutput` was maintaining internally the count of bytes written, and setting it from here would set the non-compressed value, while normally `HttpOutput` would record the compressed ones.

Also, it overwrites the value with just the last buffer length, so does it lose all the previous buffers length?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java (hunk 1 above)
Comment (joakime) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java:342]: Gotcha.
When testing the `org.eclipse.jetty.ee9.servlet.CrossContextDispatcherTest.testForwardToResourceHandler` with just the `abort(x)` change alone it seems to be happy.

The Content-Length and HttpOutput.written length efforts don't seem to be related.
What other test cases are these fixing?

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpChannel.java (hunk 2 above)
Comment (gregw) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpChannel.java:563]: That's what I'm asking you as the reviewer!)
The issue was that we have an exception whilst trying to write an error page and we handle it by trying to write an error page.... which resulted in a nasty loop.   I think aborting if an error page cannot be generated is the best thing to do

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java (hunk 3 above)
Comment (gregw) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java]: That is why we have two counts of bytes written.  The real one is down in HttpChannel and it records bytes going to the wire (after compression etc.).   The count in `HttpOutput` is an API count and is needed to implement the servlet spec behaviour of committing when the number of bytes set in content length have been written.   By its nature, it has to be before the write is attempted, as the write may be different if the count == contentLength.

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java (hunk 1 above)
Comment (gregw) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ServletCoreResponse.java:342]: The content-length change was just an improvement notice whilst debugging the comparison between the bytes written and the content-length set.   I noticed we were using the addHeader mechanism that was doing integer to string and string to integer conversion when it was not necessary.  So this change is not a fix, just an optimization.

**Comment by gregw** [comment]
This issue is somewhat related to #11325

**Review by joakime — APPROVED** [review]

**Review by sbordet — APPROVED** [review]

**Comment by joakime** [comment]
Waiting on CI green before I merge and stage 12.0.16

**Pull request merged**